docs/networking/streamcodecs.md

```diff
@@ -0,0 +1,369 @@
+
+### Map
+
+A stream codec for a map of key and value objects can be generated using two stream codecs via `ByteBufCodecs#map`. The function also takes in an `IntFunction` that constructs the empty map and an optional maximum size.
+
+```java
+public static final StreamCodec<ByteBuf, Map<String, BlockPos>> MAP_STREAM_CODEC =
+    ByteBufCodecs.map(
+        HashMap::new, // Constructs a map with the specified capacity
+        ByteBufCodecs.STRING_UTF8,
+        BlockPos.STREAM_CODEC,
+        256 // The map can only have up to 256 elements
+    );
+```
+
+### Either
+
+A stream codec for two different methods of reading/writing some object data can be generated from two steram codecs via `ByteBufCodecs#either`. This method first reads/writes a boolean indicating whether to read/write the first or second stream codec, respectively.
+
+```java
+public static final StreamCodec<ByteBuf, Either<Integer, String>> EITHER_STREAM_CODEC = 
+    ByteBufCodecs.either(
+        ByteBufCodecs.VAR_INT,
+        ByteBufCodecs.STRING_UTF8
+    );
+```
+
+### Id Mapper
+
+In most cases, when sending information across the network where an object is present on both sides, an integer representing an id is sent. Ids representing an object reduce the amount of information that need to be synced across the network. Both enums and registries make use of this.
+
+`ByteBufCodecs#idMapper` provides a convenient way to send ids for objects. It either takes in two functions which convert an object to int and vice versa, or an `IdMap`.
+
+```java
+// For some enum
+public enum ExampleIdObject {
+    ;
+
+    // Gets Id -> Enum
+    public static final IntFunction<ExampleIdObject> BY_ID = 
+        ByIdMap.continuous(
+            ExampleIdObject::getId,
+            ExampleIdObject.values(),
+        ByIdMap.OutOfBoundsStrategy.ZERO
+    );
+    
+    ExampleIdObject(int id) { /* ... */ }
+}
+
+// The stream codec would look like
+public static final StreamCodec<ByteBuf, ExampleIdObject> ID_STREAM_CODEC =
+    ByteBufCodecs.idMapper(ExampleIdObject.BY_ID, ExampleIdObject::getId);
+```
+
+:::note
+NeoForge provides an alternative for id mappers that does not cache the enum values on construction via `IExtensibleEnum#createStreamCodecForExtensibleEnum`. However, this rarely needs to be used outside of extensible enums.
+:::
+
+### Optional
+
+A stream codec for sending an `Optional` wrapped value can be generated by supplying a stream codec to `ByteBufCodecs#optional`. This method first reads/writes a boolean indicating whether to read/write the object.
+
+```java
+public static final StreamCodec<RegistryFriendlyByteBuf, Optional<DataComponentType<?>>> OPTIONAL_STREAM_CODEC =
+    DataComponentType.STREAM_CODEC.apply(ByteBufCodecs::optional);
+```
+
+### Registry Objects
+
+Registry objects can be sent across the network using one of three methods: `registry`, `holderRegistry`, or `holder`. Each takes in a `ResourceKey` representing the registry the registry object is in.
```

Comment:
This should mention that custom registries must be marked as synced for the first two to work at all and the last one to work with non-direct holders, otherwise the encoder throws an exception (added by Neo).